### ISSUE TITLE
Can we save (or speed up) a jitted Hessian computation?

### ISSUE THREAD

**Issue by rgiordan** [google/jax#4300]
We have a somewhat complicated objective function, of which we need a Hessian matrix.  The times to jit the objective function and gradient are pretty fast (about 1 and 5 seconds, respectively) but the time to jit the Hessian is much slower (about 30 seconds).  Once jitted, everything is extremely fast to evaluate (since, in our case, the Hessian is pretty sparse).  

These times are on a toy model, which we would like to scale up to larger problems.  If we pay the cost of the Hessian jit every time, our methods will be uncompetitive, but if we can cache the jitted Hessian computation, our methods will be very competitive.  (I can add more context if necessary.)

Ideally, we would be able to save the jitted Hessian computatoin to disk somehow.  I found the following issues that inconclusively discuss the possibility of doing something like this.  Is this something that's possible now (with some extra work) or in the future?  I should add that we'd be happy to stick the CPU---we don't need this to work on GPUs.

https://github.com/google/jax/issues/476
https://github.com/google/jax/issues/679
https://github.com/google/jax/issues/1566

Of course, speeding the Hessian jit up would also be a good solution.  Again, our model is probably too complicated to usefully reproduce here, but below is a toy example.  Obviously this example is fast enough for practical purposes, but it is like our problem in that (1) it takes about an order of magnitude longer to jit a Hessian than to jit the original function and (2) the Hessian is sparse because of interactions in an einsum.

```
import jax
import jax.numpy as np
import numpy as onp
import time

def time_jit(f, desc):
    print(f'\n==============\nTiming for {desc}')
    tic = time.time()
    f()
    print(f'First evaluation: {time.time() - tic}')
    
    tic = time.time()
    f()
    print(f'Second evaluation: {time.time() - tic}')

x = np.array(onp.random.random((50, 30)))
y = np.array(onp.random.random((30, 40)))

def foo(x, y):
    a = np.einsum('ij,jk->ik', x, y)
    return np.sum(a ** 2)

foo_jit = jax.jit(foo)
time_jit(lambda: foo(x, y), 'function')

foo_grad_jit = jax.jit(jax.grad(foo))
time_jit(lambda: foo_grad_jit(x, y), 'grad')

foo_hess_jit = jax.jit(jax.hessian(foo))
time_jit(lambda: foo_hess_jit(x, y), 'hessian')

```

Output:

```
==============
Timing for function
First evaluation: 0.013594627380371094
Second evaluation: 0.010229110717773438

==============
Timing for grad
First evaluation: 0.05717968940734863
Second evaluation: 0.0001628398895263672

==============
Timing for hessian
First evaluation: 0.25292301177978516
Second evaluation: 0.0005974769592285156
```

**Comment by cisprague** [comment]
I'm in the same situation — saving jitted functions like Hessians would be great! It would also be useful if we could automatically get the sparsity pattern.
